src/thread/mutex.cr

```diff
@@ -5,8 +5,14 @@ class Thread
   # :nodoc:
   class Mutex
     def initialize
-      ret = LibC.pthread_mutex_init(out @mutex, nil)
+      attributes = uninitialized LibC::PthreadMutexattrT
+      LibC.pthread_mutexattr_init(pointerof(attributes))
+      LibC.pthread_mutexattr_settype(pointerof(attributes), LibC::PTHREAD_MUTEX_ERRORCHECK)
```

Comment:
The standard/default mutex doesn't do any checks and deadlocks when a thread tries to recursively lock again. It also allows whatever thread to unlock the mutex (what?).

We could allow recursive locks but it still wouldn't prevent deadlocks, since we can miss an unlock, and it requires to keep a counter.

Erroring immediately sounds like a better solution to me.

Reply:
I.e. I believe adding error checking is required and enough. I don't foresee use-cases for recursive mutual exclusion.

Reply:
Since `Thread::Mutex` should be private and undocumented (move it under `Crystal::` eventually since user code will use `::Mutex` when it's thread-safe), the critical sections will be tiny. With `::Mutex` being non-blocking at the application level having large critical sections containing user code (which may recurse) having a recursive-save lock makes sense.


So this is fine.